File: backend/app/schemas/calculator.py
"""Hidden cost calculator request/response schemas."""
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class StateRate(BaseModel):
    """Transfer tax rate for a German state."""

    state_code: str
    state_name: str
    transfer_tax_rate: float


class CostDefaults(BaseModel):
    """Default cost percentages used in calculations."""

    notary_fee_percent: float
    land_registry_fee_percent: float
    agent_commission_percent: float


class StateRatesResponse(BaseModel):
    """Response containing all state rates and cost defaults."""

    data: list[StateRate]
    cost_defaults: CostDefaults


class HiddenCostCalculationCreate(BaseModel):
    """Request to save a hidden cost calculation."""

    name: str | None = Field(None, max_length=255)
    property_price: float = Field(..., gt=0, description="Property price in EUR")
    state_code: str = Field(..., min_length=2, max_length=2, description="German state code")
    property_type: str = Field(..., max_length=50)
    include_agent: bool = True
    renovation_level: Literal["none", "light", "medium", "full"] = "none"
    include_moving: bool = True


class HiddenCostCalculationResponse(BaseModel):
    """Full response for a saved calculation."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str | None = None
    share_id: str | None = None
    # Inputs
    property_price: float
    state_code: str
    property_type: str
    include_agent: bool
    renovation_level: str
    include_moving: bool
    # Results
    transfer_tax: float
    notary_fee: float
    land_registry_fee: float
    agent_commission: float
    renovation_estimate: float
    moving_costs: float
    total_additional_costs: float
    total_cost_of_ownership: float
    additional_cost_percentage: float
    created_at: datetime


class HiddenCostCalculationSummary(BaseModel):
    """Summary for list views."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str | None = None
    share_id: str | None = None
    property_price: float
    state_code: str
    total_additional_costs: float
    total_cost_of_ownership: float
    created_at: datetime


class HiddenCostCalculationListResponse(BaseModel):
    """Paginated list of saved calculations."""

    data: list[HiddenCostCalculationSummary]
    count: int


class StateComparisonItem(BaseModel):
    """Cost breakdown for a single state in comparison view."""

    state_code: str
    state_name: str
    transfer_tax_rate: float
    transfer_tax: float
    notary_fee: float
    land_registry_fee: float
    agent_commission: float
    total_cost: float


class StateComparisonResponse(BaseModel):
    """Response for state-by-state comparison."""

    property_price: float
    include_agent: bool
    data: list[StateComparisonItem]

File: frontend/src/models/calculator.ts
/**
 * Calculator domain models
 * TypeScript interfaces for hidden cost and ROI calculations
 */

export interface StateRate {
  stateCode: string
  stateName: string
  transferTaxRate: number
}

export interface CostDefaults {
  notaryFeePercent: number
  landRegistryFeePercent: number
  agentCommissionPercent: number
}

export interface StateRatesResponse {
  data: StateRate[]
  costDefaults: CostDefaults
}

export interface HiddenCostCalculationInput {
  name?: string
  propertyPrice: number
  stateCode: string
  propertyType: string
  includeAgent: boolean
  renovationLevel: "none" | "light" | "medium" | "full"
  includeMoving: boolean
}

export interface HiddenCostCalculation {
  id: string
  name?: string
  shareId?: string
  propertyPrice: number
  stateCode: string
  propertyType: string
  includeAgent: boolean
  renovationLevel: string
  includeMoving: boolean
  transferTax: number
  notaryFee: number
  landRegistryFee: number
  agentCommission: number
  renovationEstimate: number
  movingCosts: number
  totalAdditionalCosts: number
  totalCostOfOwnership: number
  additionalCostPercentage: number
  createdAt: string
}

export interface HiddenCostCalculationSummary {
  id: string
  name?: string
  shareId?: string
  propertyPrice: number
  stateCode: string
  totalAdditionalCosts: number
  totalCostOfOwnership: number
  createdAt: string
}

export interface StateComparisonItem {
  stateCode: string
  stateName: string
  transferTaxRate: number
  transferTax: number
  notaryFee: number
  landRegistryFee: number
  agentCommission: number
  totalCost: number
}

export interface StateComparisonResponse {
  propertyPrice: number
  includeAgent: boolean
  data: StateComparisonItem[]
}

// ---------------------------------------------------------------------------
// ROI Calculator
// ---------------------------------------------------------------------------

export interface ROICalculationInput {
  name?: string
  purchasePrice: number
  downPayment: number
  monthlyRent: number
  monthlyExpenses: number
  annualAppreciation: number
  vacancyRate: number
  mortgageRate: number
  mortgageTerm: number
}

export interface ProjectionYear {
  year: number
  propertyValue: number
  equity: number
  cumulativeCashFlow: number
  totalReturn: number
  totalReturnPercent: number
}

export interface ROICalculation {
  id: string
  name?: string
  shareId?: string
  // Inputs
  purchasePrice: number
  downPayment: number
  monthlyRent: number
  monthlyExpenses: number
  annualAppreciation: number
  vacancyRate: number
  mortgageRate: number
  mortgageTerm: number
  // Results
  grossRentalIncome: number
  netOperatingIncome: number
  annualCashFlow: number
  monthlyMortgagePayment: number
  grossYield: number
  netYield: number
  capRate: number
  cashOnCashReturn: number
  investmentGrade: number
  investmentGradeLabel: string
  // Projections
  projections: ProjectionYear[]
  createdAt: string
}

export interface ROICalculationSummary {
  id: string
  name?: string
  shareId?: string
  purchasePrice: number
  investmentGrade: number
  investmentGradeLabel: string
  annualCashFlow: number
  createdAt: string
}

// ---------------------------------------------------------------------------
// Financing Eligibility Wizard
// ---------------------------------------------------------------------------

export type EmploymentStatus =
  | "permanent"
  | "fixed_term"
  | "self_employed"
  | "freelance"
  | "civil_servant"

export type SchufaRating =
  | "excellent"
  | "good"
  | "satisfactory"
  | "adequate"
  | "poor"
  | "unknown"

export type FinancingResidencyStatus =
  | "german_citizen"
  | "eu_citizen"
  | "permanent_resident"
  | "temporary_resident"
  | "non_eu"

export interface FinancingAssessmentInput {
  name?: string
  employmentStatus: EmploymentStatus
  employmentYears: number
  monthlyNetIncome: number
  monthlyDebt: number
  availableDownPayment: number
  schufaRating: SchufaRating
  residencyStatus: FinancingResidencyStatus
}

export interface FinancingAssessment {
  id: string
  name?: string
  shareId?: string
  // Inputs
  employmentStatus: string
  employmentYears: number
  monthlyNetIncome: number
  monthlyDebt: number
  availableDownPayment: number
  schufaRating: string
  residencyStatus: string
  // Score breakdown
  employmentScore: number
  incomeRatioScore: number
  downPaymentScore: number
  schufaScore: number
  residencyScore: number
  yearsBonusScore: number
  // Results
  totalScore: number
  likelihoodLabel: string
  // Estimates
  maxLoanEstimate: number
  recommendedDownPaymentPercent: number
  expectedRateMin: number
  expectedRateMax: number
  ltvRatio: number
  // Advisory
  strengths: string[]
  improvements: string[]
  documentChecklist: string[]
  createdAt: string
}

export interface FinancingAssessmentSummary {
  id: string
  name?: string
  shareId?: string
  totalScore: number
  likelihoodLabel: string
  maxLoanEstimate: number
  createdAt: string
}
